### DISCUSSION TITLE
RasPi indi-allsky -- MQTT Broker == Bridge == MQTT Broker -- VM indi-allsky

### DISCUSSION THREAD

**Discussion by hoetzgit** [aaronwmorris/indi-allsky#1639] [Q&A] [answered]
Hello,
I apologize in advance if I have overlooked any relevant documentation on this topic (please feel free to link to it). My text is automatically translated from de to en.

My desired scenario: 
I would like to use a Raspberry P4 8GB RAM and the HQ camera as an Allsky camera. If possible, the Raspberry should only perform basic tasks for this purpose, such as image creation. The Raspberry is connected to my network via W-LAN. As far as possible, all post-processing activities for an image should then be carried out on a high-performance virtualized computer under Proxmox.
I would like the image creation and image processing to work asynchronously. This means, for example, that the Raspberry creates an image and saves it together with metadata (camera data, sensor data, weather, etc.) in an MQTT message broker. This is connected to a broker in the high-performance network via a bridge.
This would not lead to an image list even in the event of brief network problems. The VM computer then reads this data for post-processing.

My questions are:
1. can this be implemented - possibly partially - with indi-allsky?
2. if so, is there any documentation for this scenario?
3. if no documentation is available, how must the Raspberry and the VM computer be configured, i.e. which operating system is best suited in each case, which components must be installed in each case and which configuration deviating from the standard must be carried out on the Respberry and the VM computer?
4. where are the relevant configurations for image creation created?
5. where are the configurations required for post-processing created?

Unfortunately, I am not familiar with indi-specific processing. I would be very happy if I could receive support on this.

Many thanks and best regards
Henry

**Comment by aaronwmorris** [comment]
There is not really a way to do this, but I am not sure it is really needed.

If the goal is asynchronous processing, indi-allsky already performs most functions asynchronously.  Image capture occurs in a separate process from image processing (timelapse generation and file transfers also happen in separate processes/threads respectively).

If you are interested in just offloading all processing, because INDI is a full client/server architecture, you could just run the indiserver process on the Raspberry PI, which will manage the camera, then the remote indi-allsky just sends commands and receives images from the remote indiserver.  You would have to use [indi_pylibcamera](https://github.com/scriptorron/indi_pylibcamera) to utilize this mode with the HQ camera.

**Reply by hoetzgit** [reply]
@aaronwmorris Hello Aaron, thank you for your reply.
My current priority is to relieve the Raspberry of image processing. I should be able to achieve this with the method you have outlined. Later on, I want to become independent of a permanent network connection, as my WLAN connection is not always robust.

Now I have the following questions:
1. i have to install indi-allsky on both systems (RasPi/VM). How do the installations differ?
2. where are the corresponding configurations made? Is there a “master” on which all settings are made? Or do the settings for the camera have to be made on the RasPi and the settings for the post-processing on the VM? The settings for the camera are then ignored on the VM, and the settings for the post-processing are ignored on the RasPi?

Many thanks for your support
Henry

**Reply by aaronwmorris** [reply, marked as answer]
In order to use as remote indiserver, you only need the indi (and indi-3rdparty) component on the raspberry pi.  indi-allsky would not need to be installed on the raspberry pi.  There are no settings that need to be maintained on the indiserver component.  You would just need to ensure the indiserver starts automatically on the raspberry pi.

indi-allsky would only need to be installed on the VM.  All of the configuration for indi-allsky is stored in the database (sqlite by default).

When the indiserver is started, it starts listening on port 7624 and indi-allsky can connect to that port remotely.  The hostname of the indiserver connection is on the Camera tab in the config page.

**Reply by hoetzgit** [reply]
@aaronwmorris
Hello Aaron, I think your answers are a good basis for trying out this configuration.

There may be further questions, I hope I can count on your support again.

Many thanks and best regards
Henry

**Comment by HenningSchuhmann** [comment]
Hi,
I'm planing the same right now.
The VM for the image processing is up and running (INDI Passive - Camera installation).
The remote Camera is an raspberry Pi zero 2W or an Raspberry Pi3B with an RPI HQ Camera.
But how do I only install the indiserver on it? I when I choose only the indi core (without 3rd party - not needed for RPI HQ Cam). I will get the full installation... (takes a while to compile....)
And even if thats okay, do I only have to put the IP of the remote Camera inte the Box "INDI Server" of the image Processing PC?
Best Regards
Henning

**Reply by aaronwmorris** [reply]
You probably do not want to use `INDI Passive`  That is when you want to connect a second indi-allsky instance to an indiserver that is controlled by another instance of indi-allsky.

There is a script `./misc/indiserver_only_setup.sh` that is used to only setup the indiserver on the Pi.  In order to use a remote camera with an RPI HQ Camera, you will either need to setup `indi_pylibcamera` or compile the `indi_libcamera_ccd` driver from the 3rd party repo.

https://github.com/aaronwmorris/indi-allsky/wiki/indi_pylibcamera-setup  (this is for indi_pylibcamera)

Once you have one of those setup, all you should have to do is put the IP address of the PI running the indiserver in the config and start up.

**Comment by HenningSchuhmann** [comment]
Hi Aaron,

thank you so very much for your help, I really appreciate it!

On the camera machine:
I also did the` indiserver_only_setup.sh` setup, my mistake was probably to run `./setup.sh` again afterwards.
I see that `indi_pylibcamera` needs a bit more ram, which I don't have on this machine. So I would have to build `indi_libcamera_ccd`. Are there any instructions for this?

On the image processing server:
As described, I have chosen `indi passive` as I have read this from the instructions. What do I have to choose instead? Should I select the camera that I have connected remotely?

Best regards
Henning

**Reply by HenningSchuhmann** [reply]
Hi Aaron,
I've made some progress. 
What I have done (so far):

```
git clone --depth=1 https://github.com/indilib/indi-3rdparty
sudo apt-get install libboost-program-options1.74-dev
mkdir build
cd build
cmake -DCMAKE_INSTALL_PREFIX=/usr ..
make
```

that worked so far, now it seems I still have to activate it...

**Reply by aaronwmorris** [reply]
You can run the `./misc/build_indi.sh`.  INDI Core needs to be built then you can select the individual indi_libcamera_ccd driver during the 3rdparty build.

**Reply by HenningSchuhmann** [reply]
Thank you, I think I will start the install from the scratch again...

How do I have to the up my Image Processing System to conntec to the indi camera? Do I have to choose the `INDI` Camera Interface?

**Reply by aaronwmorris** [reply]
Yes, you would choose the INDI interface.  Once you are able to login to the web interface, you may then change the hostname of the indiserver.

Back to indi_pylibcamera... the Pi3 should have plenty of RAM to run just the indiserver with inid_pylibcamera.  The warning around RAM is if you try to run the indiserver on the same system as the capture system.  But 1GB should be fine if you only run the indiserver.

**Reply by HenningSchuhmann** [reply]
Okay, 
is there any recommendation: `indi_pylibcamera` or `indi_libcamera_ccd`?

**Reply by aaronwmorris** [reply]
`indi_pylibcamera` is probably more mature, but it is just a little more difficult to setup due to the manual steps.

Functionally, they should be roughly the same.

**Reply by HenningSchuhmann** [reply]
Hi Aaron,

I really appreciate your help, but I have to say that I'm about to give up.
I've only been building packages for 6 days, as this takes hours on my hardware. I feel like I'm constantly going round in circles and it's just try & error as I don't have a clear roadmap.

I had just thought I had installed everything cleanly again, but the “virtualenv/indi-allsky/pyvenv.cfg” is missing to continue...

I also made a new VM for the image processing server (luckily it's faster there) to compile it again with the INDI Camera Interface, but then in the next step it asks me for the indi camera where in my opinion there is nothing in the selection that makes sense in my case.


Sorry, is there a clear wiki somewhere? 
Can you tell me what exactly I have to set for which system during setup?

**Reply by aaronwmorris** [reply]
Building the software on the Pi3 is a painfully slow process.  The part about the virtualenv assumes the indiserver and capture process will be running on the same device.  Normally, the indiserver does not need a python virtualenv.

In this case, I would just build the indi_libcamera_ccd driver.  You do not need to rebuild the entirety of INDI, just select the 3rd party install, and only select the libcamera driver.

**Reply by HenningSchuhmann** [reply]
Okay,

thank you very much. I think I will follow this path. 
I have now:
`./misc/indiserver_only_setup.sh` executed (still without 3rd party...)
then I am asked to compile everything using 
`./misc/build_indi.sh`
In the prompt to call `build_indi.sh` it says that I should then start `setup.sh`, but it feels like I start from the beginning again...
How can I make sure that only the indi_server is installed and then also tested?

Best regards
Henning

**Reply by aaronwmorris** [reply]
For the indiserver setup on the Pi3, you should only need to run 3 things:

1. `./misc/build_indi.sh` - INDI Core & INDI 3rdparty (only indi_libcamera_ccd).  My recollection is this would probably take about 2 hours on the Pi3.
1. `./misc/indiserver_only_setup.sh` - This will ensure a running indiserver is available.

On the processing server:

1. `./misc/build_indi.sh` - Only INDI Core is needed.  If you use Ubuntu, INDI components would be installed from the INDI PPA.
1. `./setup.sh` - Just select indi_simulator_ccd for the indiserver here since you do not need it.

**Reply by HenningSchuhmann** [reply]
Hi!
Thank you for your reply. I managed to get it to work yesterday, also with your information above. However, I took a different route on the camera computer. I installed the Indi Core per package and built the indi_libcamera_ccd from 3rdparty sources. In detail:

`sudo apt update && sudo apt upgrade`
`sudo apt install -y git cmake build-essential libindi-dev indi-bin libcfitsio-dev libusb-1.0-0-dev libudev-dev libjpeg-dev libcurl4-gnutls-dev libgsl-dev libboost-dev libboost-system-dev libboost-thread-dev libcamera-dev libcamera-apps`

`git clone https://github.com/indilib/indi-3rdparty.git`
`cd indi-3rdparty/indi-libcamera`
`mkdir -p build`
`cd build`
`cmake ..`
`make -j1`
`sudo make install`
`sudo ldconfig`
Stop Indi Server:
`pkill indiserver || true`
Start indi Server with libcamera driver:
`indiserver indi_libcamera_ccd`

now I get the Images on the image processing PC

**Reply by HenningSchuhmann** [reply]
Just a quick question to confirm my assumption:

The image capture parameters I enter on the image processing computer (binning,...) are they transferred to the camera computer and then used for image capture, or do I have to configure something on the camera Raspi?

Best regards

**Reply by aaronwmorris** [reply]
Yes, all of the capture parameters are managed on the image processing side.  There is very little you have to configure on the indiserver side as all of the camera settings are managed by the capture client.

**Reply by HenningSchuhmann** [reply]
Perfect! Thanky you for your help!
